### ISSUE TITLE
Test ValidationLoop.run() also resets DataLoader before iterating

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#3688]
The `run_subset()` reset behavior is now tested, but `run()` in `validation_loop.mojo` may have the same pattern. Check whether `run()` also calls `val_loader.reset()` internally, and if so, add a parallel test `test_validation_loop_run_resets_loader` using the same pre-exhaustion strategy (set `current_batch = num_batches` before calling `run()`).

_Follow-up from #3186_

**Comment by mvillmow** [comment]
# Implementation Plan

I now have all the context I need. Here's the implementation plan:

---

## Implementation Plan: Issue #3688 — Test `ValidationLoop.run()` Resets DataLoader

### 1. Objective

Add `test_validation_loop_run_resets_loader` to `tests/shared/training/test_validation_loop.mojo` to verify that `ValidationLoop.run()` resets a pre-exhausted `DataLoader` before iterating, parallel to the existing `test_validation_loop_run_subset_resets_loader`.

### 2. Approach

**`run()` does reset the loader** — confirmed at line 94 of `validation_loop.mojo`, inside `validate()` which `run()` delegates to. Additionally, `run()` calls `val_loader.reset()` a second time at line 218 for its accuracy pass. Both resets are covered by the same pre-exhaustion strategy.

Strategy mirrors `test_validation_loop_run_subset_resets_loader` exactly:

1. Create a 2-batch loader
2. Pre-exhaust it (`loader.current_batch = loader.num_batches`)
3. Assert `not loader.has_next()` to confirm pre-condition
4. Call `vloop.run()` — which internally calls `validate()` which calls `val_loader.reset()`
5. Assert the returned loss is a valid finite value (not NaN/division-by-zero)

The self-proving failure mode: if `reset()` were absent, `validate()` would process 0 batches → `total_loss / Float64(0)` → NaN or exception. A valid finite loss proves reset occurred.

### 3. Files to Create

None — no new files needed.

### 4. Files to Modify

**`tests/shared/training/test_validation_loop.mojo`**

- **Add** `test_validation_loop_run_resets_loader()` function after the existing `run_subset` reset test (around line 271)
- **Add** call to `test_validation_loop_run_resets_loader()` in `main()` under the `"Running ValidationLoop.run() tests..."` block (after line 477)

### 5. Implementation Order

1. **Read** the full test file to confirm exact insertion points (already done)
2. **Add** the new test function after `test_validation_loop_run_subset_resets_loader` (line 270):

```mojo
fn test_validation_loop_run_resets_loader() raises:
    """Test run() resets a pre-exhausted DataLoader before iterating.

    Strategy: Create a loader with exactly 2 batches, then exhaust it by
    setting current_batch = num_batches. Without reset(), has_next() returns
    False immediately -> 0 batches processed -> division by zero. With reset(),
    the loader restarts and processes all 2 batches -> valid finite loss.

    This proves run() calls val_loader.reset() internally (via validate(),
    line 94 of validation_loop.mojo).
    """
    var vloop = ValidationLoop()
    # 2 batches total (8 samples, batch_size=4)
    var loader = create_val_loader(n_batches=2)
    # Pre-exhaust: advance to end so has_next() returns False
    loader.current_batch = loader.num_batches
    assert_true(not loader.has_next())
    var metrics = TrainingMetrics()
    # run() delegates to validate() which calls val_loader.reset() internally
    var val_loss = vloop.run(simple_forward, simple_loss, loader, metrics)
    # Valid finite loss proves batches were processed after reset (not 0)
    assert_greater(val_loss, Float64(-1e-10))
    assert_less(val_loss, Float64(1e10))
    print("  test_validation_loop_run_resets_loader: PASSED")
```

3. **Update** `main()` — add call in the `run()` tests block:

```mojo
    print("Running ValidationLoop.run() tests...")
    test_validation_loop_run_basic()
    test_validation_loop_run_updates_metrics()
    test_validation_loop_run_compute_accuracy_false()
    test_validation_loop_run_resets_loader()   # ← add this
```

### 6. Verification

- **Local**: Cannot run `mojo test` locally due to GLIBC incompatibility — use CI or Docker
- **CI**: Push branch, CI runs `just test-mojo` or `just test-group tests/shared/training "test_*.mojo"` — confirm the new test appears in output as `PASSED`
- **Failure mode check**: Temporarily remove the `val_loader.reset()` call from `validate()` (line 94) and confirm the test fails with NaN/exception — this validates the test is actually proving the behavior, not trivially passing

### 7. Skills Used

**During planning (invoked):**
- `docs-search` / `Explore` agent — explored codebase to locate `validation_loop.mojo`, `test_validation_loop.mojo`, and `DataLoader` struct

**From team knowledge base (Prior Learnings):**
- `dataloader-reset-verification` — provided the pre-exhaustion strategy (`loader.current_batch = loader.num_batches`), self-proving failure mode design, and range assertion bounds
- `mojo-validation-loop-accuracy-tracking` — confirmed `run()` internals and the second-pass reset pattern at line 218, and warned against tuple return syntax

---
*Generated by Claude Code Planner*
